R/descriptives.R

```diff
@@ -1239,6 +1289,111 @@ Descriptives <- function(jaspResults, dataset, options) {
   return(p)
 }
 
+
+
+.plotMarginal_Overlay <- function(column, variableName, overlayFactor, splitByName,
+                          rugs = FALSE, displayDensity = FALSE, binWidthType = c("doane", "fd", "scott", "sturges", "manual"),
+                          numberOfBins = NA,
+                          lwd = 1) {
+  binWidthType <- match.arg(binWidthType)
+  #column <- as.numeric(column)
+  variable <- column[[variableName]] #na.omit(column)
+
+  #message(variable)
+
+  if (length(variable) == 0)
+    return(NULL)
+
+  if (binWidthType == "doane") {  # https://en.wikipedia.org/wiki/Histogram#Doane's_formula
+    sigma.g1 <- sqrt((6*(length(variable) - 2)) / ((length(variable) + 1)*(length(variable) + 3)))
+    g1 <- mean(abs(variable)^3)
+    k <- 1 + log2(length(variable)) + log2(1 + (g1 / sigma.g1))
+    binWidthType <- k
+  } else if (binWidthType == "fd" && nclass.FD(variable) > 10000) { # FD-method will produce extreme number of bins and crash ggplot, mention this in footnote
+    binWidthType <- 10000
+  } else if (binWidthType == "manual") {
+    binWidthType <- numberOfBins
+  }
+
+
+  h <- hist(variable, plot = FALSE, breaks = binWidthType)
+
+  if (!displayDensity)
+    yhigh <- max(h$counts)
+  else {
+    dens <- density(variable)
+    yhigh <- max(max(h$density), max(dens$y))
+  }
+
+  ylow <- 0
+  xticks <- base::pretty(c(variable, h$breaks), min.n = 3)
+
+  if (!displayDensity) {
+    p <-
+      jaspGraphs::drawAxis(
+        xName = variableName, yName = gettext("Counts"), xBreaks = xticks,
+        yBreaks = base::pretty(c(0, h$counts)), force = TRUE, xLabels = xticks
+      )
+  } else {
+    p <-
+      jaspGraphs::drawAxis(
+        xName = variableName, yName = gettext("Density"), xBreaks = xticks,
+        yBreaks = c(0,  1.05 * yhigh), force = TRUE, yLabels = NULL,
+        xLabels = xticks
+      )
+  }
+
+
+  if (displayDensity) {
+    p <- p +
+      ggplot2::geom_histogram(
+        data    = data.frame(x = variable, g = overlayFactor),
+        mapping = ggplot2::aes(x = x, fill=factor(g), y = ..density..),#options$splitBy, y = ..density..),
```

Comment:
Perhaps it makes sense to incorporate these changes [here](https://github.com/jasp-stats/jaspGraphs/pull/57)? If you prefer, I can also do that (either before or after this PR).

Reply:
Sounds good, I have a very slight preference for moving onto something else in the interest of getting broader exposure to different parts of JASP, but emphasis on the word 'slight' - I would also happily have a go at incorporating the changes in jaspGraphs. By the way I think it's worth mentioning that there were a couple of not-completely-resolved questions about whether this could be implemented more neatly, and what to do if there is no split factor. Are there differences between the required functionality or code standards in the jaspGraphs and jaspDescriptives modules?

Reply:
> I have a very slight preference for moving onto something else in the interest of getting broader exposure to different parts of JASP

Makes sense.

> By the way I think it's worth mentioning that there were a couple of not-completely-resolved questions about whether this could be implemented more neatly, and what to do if there is no split factor.

This can definitely be done in one function (and should probably be done in one function).
Personally, I'd implement this somewhat like this:

```r
.plotMarginal_Overlay <- function(column, variableName, overlayFactor = NULL, splitByName = NULL, ...) {

if (displayDensity) {
  if (is.null(overlayFactor)) {
    histogramData    <- data.frame(x = variable, g = overlayFactor)
    histogramMapping <- ggplot2::aes(x = x, fill=factor(g), y = ..density..)
  } else {
    histogramData    <- data.frame(x = variable)
    histogramMapping <- ggplot2::aes(x = x, y = ..density..)
  }

  p <- p +
    ggplot2::geom_histogram(
      data    = histogramData,
      mapping = histogramMapping,
      ...
    ) + ...
}
```

> Are there differences between the required functionality or code standards in the jaspGraphs and jaspDescriptives modules?

Since jaspGraphs is reused by all of JASP, I tend to be a bit more strict in terms of functionality and documentation (e.g., enforcing roxygen2 for documentation). Also, all code there must pass the R CMD CHECK. but I can also do that if you prefer to work on other things!